authorization/permissions.py

```diff
@@ -148,25 +167,26 @@ def has_permission(self, request, view):
 # ==================
 
 
-class ObjectVisibleBasePermission(MessageMixin, Permission):
-    model = None
-    obj_var = None
+TModel = TypeVar('TModel')
+class ObjectVisibleBasePermission(MessageMixin, Permission, Generic[TModel]):
+    model: Type[TModel]
+    obj_var: str
 
-    def has_permission(self, request, view):
+    def has_permission(self, request: HttpRequest, view: Any):
         obj = getattr(view, self.obj_var, None)
         return (
             obj is None or
             self.has_object_permission(request, view, obj)
         )
 
-    def has_object_permission(self, request, view, obj):
-        user = request.user
+    def has_object_permission(self, request: HttpRequest, view: Any, obj: TModel) -> bool:
+        user = cast(User, request.user)
         return (
             not isinstance(obj, self.model) or # skip objects that are not model in question
```

Comment:
This should be fixed for the other has_object_permission methods as well.

Reply:
Yes, I agree that in a fully annotated codebase (with no type errors) the isinstance check should always be true. However, most of the codebase is still unannotated, and I thought that there *might* be some odd cases where an object of a different type is passed to the method. Because of this, I'm not sure if we should remove the check yet since it might have some unintended consequences.

Reply:
> I think it should be the former as has_object_permission seems to be designed to take any model type.

That is, leave the isinstance check and replace TModel with models.Model. The "The type should be the one that the design dictates (as opposed to what the usage dictates)" rule in the orientation specifies that it should be a Model as that method seems to be designed to take any model and see if the permission disallows it

Reply:
The reason I think something should be done is that either the type hint lies or the check is redundant. Neither is good. Simplest fix that also seems to coincide with the original design is to do what I just posted.

Reply:
Oh now I get what you mean. Yes, the method seems to be designed to take any model instance (hence the isinstance check), which means the type hint should be Model instead of the more specific type TModel (and as a consequence, the class does not need to be generic). Good catch, I'll fix that.

Reply:
The is_object_visible method is designed to only take TModel, so I think it should be generic

Reply:
That's correct. I forgot there was yet another method called after has_object_permission. So yes, the class shall remain generic.

Reply:
I think that the `isinstance` exists in the old code because there are really some cases when `obj` has a different type and the check prevents it from crashing.